Dockerfile

```diff
@@ -5,13 +5,14 @@ ENV PATH="/PrairieLearn/node_modules/.bin:$PATH"
 # Install Python/NodeJS dependencies before copying code to limit download size
 # when code changes.
 #
-# Note that we also have to copy the `packages` directory so that `yarn`
-# can resolve the workspaces inside it and set up symlinks correctly.
-# This is suboptimal, as a change to any file under `package/` will
+# Note that we also have to copy the `apps` and `packages` directories so that
+# `yarn` can resolve the workspaces inside it and set up symlinks correctly.
+# This is suboptimal, as a change to any file in these directories will
 # invalidate this layer's cache, but it's unavoidable.
```

Comment:
I also can't think of any other way around it.

However, I just want to make sure that we really should be so worried about layer caching. What's the use case here? People doing local dev? Afaik, our executor image won't benefit from layer caching because we never pull updated versions?

Assuming we COPY `packages/package-1/package.json`, etc, then `apps/app-1/package.json`, I can see three common cases:
1. All `package.json` files are unmodified: big win from caching.
2. The PL `package.json` has a change, but all of the `packages/` have not changed. Some win from caching, but I'm not sure how much of a win this actually is? PL presumably has by far the largest `node_modules`?
3. One of the first COPYed `packages/` has a change in `package.json`: almost all caches are blown away.

So I guess I'm wondering just much win we get in case (2), which I think is quite common, and how often we get case (1)? Unless people are pulling new images very frequently, I think they are likely to normally hit case (2). Certainly if they pull once a month they are guaranteed to hit (2) because of dependabot.

Am I thinking about this correctly?

Reply:
The main thing I'm thinking about is CI times. For a typical master build that changes code but not dependencies, copying the entire `apps/` directory in one go means that we'll incur the cost of downloading + building all deps _and_ the cost of pushing new layers to Docker Hub. That may be several minutes, which isn't a ton, but I'm already struggling to keep our time from merge to deploy low, and this just makes that worse.

Docker appears to be completely unwilling to allow for reasonable glob behavior for `COPY`: https://github.com/moby/moby/issues/15858.

Reply:
I took the individual copy approach in f402ab6, which seems to work as expected. I added `tools/validate-dockerfile.mjs` to make sure we don't forget to add any packages.

Reply:
Ugh, it inexplicably takes 2.5 minutes to do all the `COPY` steps on GitHub Actions??

Reply:
I'm trying the "copy all of `apps/` and `packages/`" approach in 784ec9e, we'll see how that fares.

Reply:
It's marginally faster in the case of a cache miss, at the cost of incurring many more cache misses.

I think we just need to hope that Docker implements `COPY --parents` soon (https://github.com/moby/moby/issues/35639) 😢

Reply:
Ok, I agree this is all we can do for now. And thanks for reminding me about CI being the main use case for this! I always manage to forget about CI 😅